pkg/prober/prober.go

```diff
@@ -0,0 +1,150 @@
+}
+
+// RegisterInRouter registers readiness and liveness probes to mux.
+func (p *Prober) RegisterInRouter(router *route.Router) {
+	router.Get(healthyEndpointPath, p.probeHandlerFunc(p.IsHealthy, "healthy"))
+	router.Get(readyEndpointPath, p.probeHandlerFunc(p.IsReady, "ready"))
+}
+
+// RegisterInMux registers readiness and liveness probes to mux.
+func (p *Prober) RegisterInMux(mux *http.ServeMux) {
+	mux.HandleFunc(healthyEndpointPath, p.probeHandlerFunc(p.IsHealthy, "healthy"))
+	mux.HandleFunc(readyEndpointPath, p.probeHandlerFunc(p.IsReady, "ready"))
+}
+
+func (p *Prober) writeResponse(w http.ResponseWriter, probeFunc func() error, probeType string) {
+	err := probeFunc()
+	if err != nil {
+		http.Error(w, fmt.Sprintf(errorProbeFmt, p.getComponent(), probeType, err), probeErrorHTTPStatus)
+		return
+	}
+	if _, e := io.WriteString(w, fmt.Sprintf(okProbeFmt, p.getComponent(), probeType)); e == nil {
+		level.Error(p.logger).Log("msg", "failed to write probe response", "probe type", probeType, "err", err)
+	}
+}
+
+func (p *Prober) probeHandlerFunc(probeFunc func() error, probeType string) func(http.ResponseWriter, *http.Request) {
+	return func(w http.ResponseWriter, _ *http.Request) {
+		p.writeResponse(w, probeFunc, probeType)
+	}
+}
+
+// IsReady returns error if component is not ready and nil if it is.
+func (p *Prober) IsReady() error {
+	p.readyMtx.Lock()
+	defer p.readyMtx.Unlock()
+	return p.readiness
+}
+
+// SetReady sets components status to ready.
+func (p *Prober) SetReady() {
+	if p.IsReady() != nil {
+		level.Info(p.logger).Log("msg", "changing probe status", "status", "ready")
+	}
+	p.SetNotReady(nil)
+}
+
+// SetNotReady sets components status to not ready with given error as a cause.
+func (p *Prober) SetNotReady(err error) {
+	if err != nil && p.IsReady() == nil {
+		level.Warn(p.logger).Log("msg", "changing probe status", "status", "not-ready", "reason", err)
+	}
+	p.readyMtx.Lock()
+	defer p.readyMtx.Unlock()
+	p.readiness = err
+}
+
+// IsHealthy returns error if component is not healthy and nil if it is.
+func (p *Prober) IsHealthy() error {
+	p.healthyMtx.Lock()
+	defer p.healthyMtx.Unlock()
+	return p.healthiness
+}
+
+// SetHealthy sets components status to healthy.
+func (p *Prober) SetHealthy() {
+	if p.IsHealthy() != nil {
+		level.Info(p.logger).Log("msg", "changing probe status", "status", "healthy")
+	}
+	p.SetNotHealthy(nil)
+}
+
+// SetNotHealthy sets components status to not healthy with given error as a cause.
+func (p *Prober) SetNotHealthy(err error) {
+	if err != nil && p.IsHealthy() == nil {
+		level.Warn(p.logger).Log("msg", "changing probe status", "status", "unhealthy", "reason", err)
+	}
+	p.healthyMtx.Lock()
+	defer p.healthyMtx.Unlock()
+	p.healthiness = err
+}
+
+// HandleIfReady if probe is ready calls the function otherwise returns 503.
+func (p *Prober) HandleIfReady(f http.HandlerFunc) http.HandlerFunc {
+	return func(w http.ResponseWriter, r *http.Request) {
+		p.readyMtx.Lock()
```

Comment:
Why not reusing `p.IsReady`? (: In this case it's fine.

Reply:
Especially that we don't know how long `f` is and what it does from this function perspecitve.

Reply:
Good catch. This could cause serious issues.